tidy3d/components/data/near2far_radiation_vectors.py

```diff
@@ -485,345 +492,169 @@ def integrate_for_one_theta(i_th: int):
-        Ht_array = -Ep_array / eta
-        Hp_array = Et_array / eta
-        Hr_array = np.zeros_like(Hp_array)
-
-        dims = ("r", "theta", "phi")
-        coords = {"r": [r], "theta": theta, "phi": phi}
-
-        Er = xr.DataArray(data=Er_array[None, ...], coords=coords, dims=dims)
-        Et = xr.DataArray(data=Et_array[None, ...], coords=coords, dims=dims)
-        Ep = xr.DataArray(data=Ep_array[None, ...], coords=coords, dims=dims)
+            for idx_f, frequency in enumerate(freqs):
+                _N_th, _N_ph, _L_th, _L_ph = self._radiation_vectors_for_surface(
+                    frequency, theta, phi, surface, self.currents[surface.monitor.name]
+                )
+                N_theta[..., idx_f] += _N_th
+                N_phi[..., idx_f] += _N_ph
+                L_theta[..., idx_f] += _L_th
+                L_phi[..., idx_f] += _L_ph
 
-        Hr = xr.DataArray(data=Hr_array[None, ...], coords=coords, dims=dims)
-        Ht = xr.DataArray(data=Ht_array[None, ...], coords=coords, dims=dims)
-        Hp = xr.DataArray(data=Hp_array[None, ...], coords=coords, dims=dims)
+        coords = {"theta": theta, "phi": phi, "f": freqs}
+        fields = {}
+        fields["Ntheta"] = Near2FarAngleDataArray(N_theta, coords=coords)
+        fields["Nphi"] = Near2FarAngleDataArray(N_phi, coords=coords)
+        fields["Ltheta"] = Near2FarAngleDataArray(L_theta, coords=coords)
+        fields["Lphi"] = Near2FarAngleDataArray(L_phi, coords=coords)
 
-        return xr.Dataset(
-            {"E_r": Er, "E_theta": Et, "E_phi": Ep, "H_r": Hr, "H_theta": Ht, "H_phi": Hp}
-        )
+        return Near2FarAngleData(monitor=monitor, **fields)
 
-    def fields_cartesian(self, x: ArrayLikeN2F, y: ArrayLikeN2F, z: ArrayLikeN2F) -> xr.Dataset:
-        """Get fields at a point relative to monitor center in cartesian coordinates.
+    def radiation_vectors_cartesian(self, monitor: Near2FarCartesianMonitor):
+        """Compute radiation vectors on a Cartesian grid in spherical coordinates.
 
         Parameters
         ----------
-        x : Union[float, Tuple[float, ...], np.ndarray]
-            (micron) x positions relative to the local origin.
-        y : Union[float, Tuple[float, ...], np.ndarray]
-            (micron) y positions relative to the local origin.
-        z : Union[float, Tuple[float, ...], np.ndarray]
-            (micron) z positions relative to the local origin.
+        monitor : :class:`.Near2FarCartesianMonitor`
+            Instance of :class:`.Near2FarCartesianMonitor` defining the far field observation grid.
 
         Returns
         -------
-        xarray.Dataset
-            xarray dataset containing (Ex, Ey, Ez), (Hx, Hy, Hz) in cartesian coordinates.
+        :class:.`Near2FarCartesianData`
+            Data structure with ``N_theta``, ``N_phi``, ``L_theta``, ``L_phi`` radiation vectors.
         """
 
+        freqs = np.atleast_1d(self.frequencies)
+        if monitor.plane_axis == 0:
```

Comment:
see my earlier comments on unpop axis. some of this repeated code could also potentially be refactored into their own methods.